File: leetcode/20 - validparenthesis/valid-parenthesis.py
class ValidParenthesis:
  def __init__(self,parenthesis):
    self.parenthesis = parenthesis
    self.symbol = { ')': '(', ']': '[', '}': '{' }
    self.stack = []

  def isValid(self):
    for character in self.parenthesis:
        if(character == '(' or character == '{' or character == '[' ):
            self.stack.append(character)
            
        if(character == ')' or character == '}' or character == ']' ):
            if (len(self.stack) == 0 or self.stack[len(self.stack) - 1] != self.symbol[character]):
                # print(self.symbol[character])
                return False;
          

            self.stack.pop()

    return len(self.stack) == 0
    
validParenthesisTrue = ValidParenthesis("()[]{}")
validParenthesisFalse = ValidParenthesis("([)]")

print(validParenthesisTrue.isValid()) #True
print(validParenthesisFalse.isValid()) #False

File: leetcode/20 - validparenthesis/valid-parenthesis.java
package Algoritms;

import java.util.ArrayDeque;
import java.util.Deque;

public class ValidParenthesis {

    public static void main(String[] args) {

        System.out.println(isValid("()[]{}")); // true
        System.out.println(isValid("([)]"));   // false
        System.out.println(isValid("([{}])")); // true
    }

    private static boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();

        for (int i = 0; i < s.length(); i++) {
            char item = s.charAt(i);

            if (item == '(' || item == '{' || item == '[') {
                stack.push(item);
                continue;
            }

            if (stack.isEmpty()) {
                return false;
            }

            char valid = stack.pop();
            switch (item) {
                case ')':
                    if (valid == '{' || valid == '[') return false;
                    break;
                case '}':
                    if (valid == '(' || valid == '[') return false;
                    break;
                case ']':
                    if (valid == '(' || valid == '{') return false;
                    break;
            }
        }

        return stack.isEmpty();
    }
}
